packages/react-dom/src/client/ReactDOM.js

```diff
@@ -758,6 +758,10 @@ ReactDOM.unstable_createRoot = function createRoot(
   container: DOMContainer,
   options?: RootOptions,
 ): ReactRoot {
+  invariant(
+    isValidContainer(container),
+    'unstable_createRoot(...): Target container is not a DOM element.',
+  );
```

Comment:
Could this be a dev `waring` instead?

Reply:
Why? It's never going to work correctly, and making it an invariant is consistent with what `ReactDOM.render` always did.

Reply:
It's also (hopefully) not a hot code path.

Reply:
Not sure if there's much point in discussing it after the PR's been merged 😄 but as for "why" – to avoid impacting the size of the production bundle. It would still fail in both cases, but with a useful message in DEV.

Reply:
Agree this should be a warning, but don't care because it's unstable

Reply:
Is size the only rationale for it being a warning?

I tend to prefer errors when, if not true, the code would later error anyway with a more obscure message.  This prevents confusing stack traces in production crash reports. Instead of 5 different potential places where it fails later with errors like `Cannot call toLowerCase of undefined`, we only get one with its own error code.

I think the upside of having a clear error code here > the downside of adding five bytes.

Reply:
I think the idea is that invariants add up, so as a general rule– we should make them DEV-only when possible.

I think it's highly likely you'd try this first in DEV mode, see the error, then fix it– so you'd get a clear error message regardless.

Reply:
When is it impossible to make a DEV-only warning, and we *have to* use an invariant?

Not trolling — I'm just not sure I understand the decision procedure you're thinking of. Or do you mean we shouldn't add new invariants at all (which could also be plausible)?

Reply:
I don't understand your question, but I'll try to rephrase. My understanding was that for things like this– where React will error in either case– it's slightly preferable to add a helpful error message in DEV mode, and let PROD just error without a (helpful) message.